File: software_challenge_sdk/src/framework/sc/shared/GameResult.java
package sc.shared;

import java.util.LinkedList;
import java.util.List;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamImplicit;

import sc.framework.plugins.SimplePlayer;
import sc.helpers.StringHelper;

@XStreamAlias(value = "result")
public class GameResult {
	private final ScoreDefinition definition;

	@XStreamImplicit(itemFieldName = "score")
	private final List<PlayerScore> scores;
	
	@XStreamImplicit(itemFieldName = "winner")
	private List<SimplePlayer> winners;

        /**
         * might be needed by XStream
         */
        public GameResult() {
            definition = null;
            scores = null;
            winners = null;
        }

	public GameResult(ScoreDefinition definition, List<PlayerScore> scores, List<SimplePlayer> winners) {
		this.definition = definition;
		this.scores = scores;
		this.winners = winners;
	}

	public ScoreDefinition getDefinition() {
		return this.definition;
	}

	public List<PlayerScore> getScores() {
		return this.scores;
	}
	
	public List<SimplePlayer> getWinners() {
		if(this.winners == null) {
			this.winners = new LinkedList<SimplePlayer>();
		} 
		return this.winners;
	}

	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder();

		int playerIndex = 0;
		for (PlayerScore score : this.scores) {
			builder.append("--------------------------\n");
			builder.append(" Player " + playerIndex + "\n");
			builder.append("--------------------------\n");
			String[] scoreParts = score.toStrings();
			for (int i = 0; i < scoreParts.length; i++) {
				String key = StringHelper.pad(this.definition.get(i).getName(),
						10);
				String value = scoreParts[i];
				builder.append(" " + key + " = " + value + "\n");
			}
			playerIndex++;
		}

		return builder.toString();
	}

	public boolean isRegular() {
		for (PlayerScore score : this.scores) {
			if (score.getCause() != ScoreCause.REGULAR) {
				return false;
			}
		}

		return true;
	}
}

File: sdk/src/framework/sc/shared/GameResult.kt
package sc.shared

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamImplicit
import sc.framework.plugins.Player
import sc.protocol.room.ObservableRoomMessage
import sc.protocol.room.RoomOrchestrationMessage

/**
 * Das Endergebnis eines Spiels.
 * @property scores die Punktzahlen aller Spieler.
 * @property winners eine Liste an Gewinnern, sofern es welche gibt.
 */
@XStreamAlias(value = "result")
data class GameResult(
        val definition: ScoreDefinition,
        @XStreamImplicit(itemFieldName = "score")
        val scores: List<PlayerScore>,
        @XStreamImplicit(itemFieldName = "winner")
        val winners: List<Player>?
): RoomOrchestrationMessage, ObservableRoomMessage {
    
    val isRegular: Boolean
        get() = scores.all { it.cause == ScoreCause.REGULAR }
    
    override fun toString() =
            "GameResult(winner=$winners, scores=[${scores.withIndex().joinToString { "Player${it.index + 1}${it.value.toString(definition).removePrefix("PlayerScore")}" }}])"
    
    override fun equals(other: Any?) =
            other is GameResult &&
                    definition == other.definition &&
                    scores == other.scores &&
                    (winners == other.winners ||
                            (winners.isNullOrEmpty() && other.winners.isNullOrEmpty()))
    
    override fun hashCode(): Int {
        var result = definition.hashCode()
        result = 31 * result + scores.hashCode()
        result = 31 * result + (winners ?: emptyList()).hashCode()
        return result
    }
    
}